SU2_CFD/src/numerics_direct_mean.cpp

```diff
@@ -327,1435 +468,1013 @@ void CCentJST_KE_Flow::ComputeResidual(su2double *val_residual, su2double **val_
-        Proj_ModJac_Tensor_ij = 0.0;
-        /*--- Compute |Proj_ModJac_Tensor| = P x |Lambda| x inverse P ---*/
-        for (kVar = 0; kVar < nVar; kVar++)
-          Proj_ModJac_Tensor_ij += P_Tensor[iVar][kVar]*fabs(Lambda[kVar])*invP_Tensor[kVar][jVar];
-        val_Jacobian_i[iVar][jVar] += 0.5*Proj_ModJac_Tensor_ij*Area;
-        val_Jacobian_j[iVar][jVar] -= 0.5*Proj_ModJac_Tensor_ij*Area;
-      }
-    }
-  }
-}
+  AD::StartPreacc();
+  AD::SetPreaccIn(Normal, nDim);
+  AD::SetPreaccIn(V_i, nDim+4);
+  AD::SetPreaccIn(V_j, nDim+4);
 
-CUpwSLAU2_Flow::CUpwSLAU2_Flow(unsigned short val_nDim, unsigned short val_nVar, CConfig *config, bool val_low_dissipation) : CNumerics(val_nDim, val_nVar, config) {
-  
-  implicit = (config->GetKind_TimeIntScheme_Flow() == EULER_IMPLICIT);
-  
-  Gamma = config->GetGamma();
-  Gamma_Minus_One = Gamma - 1.0;
-  
-  slau_low_dissipation = val_low_dissipation;
-  
-  Diff_U = new su2double [nVar];
-  Velocity_i = new su2double [nDim];
-  Velocity_j = new su2double [nDim];
-  RoeVelocity = new su2double [nDim];
-  delta_vel  = new su2double [nDim];
-  delta_wave = new su2double [nVar];
-  ProjFlux_i = new su2double [nVar];
-  ProjFlux_j = new su2double [nVar];
-  Lambda = new su2double [nVar];
-  Epsilon = new su2double [nVar];
-  P_Tensor = new su2double* [nVar];
-  invP_Tensor = new su2double* [nVar];
-  for (iVar = 0; iVar < nVar; iVar++) {
-    P_Tensor[iVar] = new su2double [nVar];
-    invP_Tensor[iVar] = new su2double [nVar];
-  }
-}
+  /*--- Variables for the general form and primitives for mass flux and pressure calculation.  ---*/
+  /*--- F_{1/2} = ||A|| ( 0.5 * mdot * (psi_i+psi_j) - 0.5 * |mdot| * (psi_i-psi_j) + N * pf ) ---*/
 
-CUpwSLAU2_Flow::~CUpwSLAU2_Flow(void) {
-  
-  delete [] Diff_U;
-  delete [] Velocity_i;
-  delete [] Velocity_j;
-  delete [] RoeVelocity;
-  delete [] delta_vel;
-  delete [] delta_wave;
-  delete [] ProjFlux_i;
-  delete [] ProjFlux_j;
-  delete [] Lambda;
-  delete [] Epsilon;
-  for (iVar = 0; iVar < nVar; iVar++) {
-    delete [] P_Tensor[iVar];
-    delete [] invP_Tensor[iVar];
-  }
-  delete [] P_Tensor;
-  delete [] invP_Tensor;
-  
-}
+  psi_i[0] = 1.0;  psi_j[0] = 1.0;
+
+  for (iDim = 0; iDim < nDim; iDim++) {
+    /*--- Velocities ---*/
+    Velocity_i[iDim] = psi_i[iDim+1] = V_i[iDim+1];
+    Velocity_j[iDim] = psi_j[iDim+1] = V_j[iDim+1];
+  }
+
+  /*--- Pressure and Density ---*/
+  Pressure_i = V_i[nDim+1];  Pressure_j = V_j[nDim+1];
+  Density_i  = V_i[nDim+2];  Density_j  = V_j[nDim+2];
+
+  /*--- Enthalpy ---*/
+  Enthalpy_i = psi_i[nVar-1] = V_i[nDim+3];
+  Enthalpy_j = psi_j[nVar-1] = V_j[nDim+3];
 
-void CUpwSLAU2_Flow::ComputeResidual(su2double *val_residual, su2double **val_Jacobian_i, su2double **val_Jacobian_j, CConfig *config) {
-   
   /*--- Face area (norm or the normal vector) ---*/
+
   Area = 0.0;
   for (iDim = 0; iDim < nDim; iDim++)
     Area += Normal[iDim]*Normal[iDim];
   Area = sqrt(Area);
-  
+
   /*-- Unit Normal 
```

Comment:
@aeroamit I did and it was more troublesome than I expected. The differentiation of the mass flux and pressure start here and end in line 1299, unfortunately git is quite confused with the changes, look only at the green chunks.
I modified the "framework" such that derived classes can set the derivatives of mdot and pressure with respect to the velocities, pressures, densities, and enthalpies. By doing so the finite differences are bypassed.

If you have those quantities worked out for the other schemes feel free to push to this PR, I am not planning to do the other schemes manually :)

Reply:
Hi @pcarruscag ,

Things look good here (+bonus fruitful learning). I have worked out the Jacobians with respect to conservative variables.  I will do the needful and take a while to push to the online repository (some work and local formalities etc.)

Cheers
Amit